### PR TITLE
feat: add optional debug logs

### PR DESCRIPTION
**Pull request by vancouverwill**
#31 

WIP

todo:
* log other endpoints
* update readme to show how to use debugger

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -288,7 +289,7 @@ class Client {
     let hint = params.id_token_hint;
 
     if (hint instanceof TokenSet) {
-      assert(hint.id_token, 'id_token not present in TokenSet');
+      assert(hint.id_token, 'endSessionUrl() id_token not present in TokenSet');
```
Comment (vancouverwill) [inline lib/client.js]: updated the error message here so it's source was transparent. I have done the same for other places where id_token is not set

**Comment by panva** [comment]
I’m not fond of the logger option, sorry. “debug” module at most

**Comment by vancouverwill** [comment]
> I’m not fond of the logger option, sorry. “debug” module at most

Thanks for looking and fair enough. Would you mind explaining why you don't like the logger option? **debug** library only supports one style of log format and we need json formatted logs in production for log search.
I looked and thought about using debug as the only logger for a while this weekend as we do need some transparency around what is going on within **node-openid-client** but using debug is an rigid approach to logs and will only suit certain use cases. This seemed a good compromise, to default to **debug** but to allow clients of **node-openid-client** to use their own logging system.

**Comment by vancouverwill** [comment]
> First, debug capabilities must not be confused with logging, i believe none of the request/response objects should ever reach a log entry unmodified, there's PII, bearer tokens and client secret credentials all over the place.

Yeah good point, the logs I put in would certainly not be safe in a production environment. Creating events would be a lotpre flexible solution for most people.
> id_token not present in TokenSet if the IdP you're using isn't consistent with returning ID Tokens i.e. after refreshes, they ought to let you know and you shouldn't blindly call tokenset claims. It is, afterall, a really stupid helper that you can do yourself IF you have an id token.
The reason I updated these error messages is each one had sent content so made it hard to see error came from. Unique errors do help, and in authorize callback the grant and token validation errors are all caught with same catch so still bit ambiguous as to what happened. What do you think about an error getting thrown in the grant request  so you at least know that is what went wrong ?
Thanks for your time 👌

**Comment by panva** [comment]
> The reason I updated these error messages is each one had sent content so made it hard to see error came from. Unique errors do help, and in authorize callback the grant and token validation errors are all caught with same catch so still bit ambiguous as to what happened. What do you think about an error getting thrown in the grant request so you at least know that is what went wrong ?

I'm sorry I don't follow, the problem is?

**Comment by vancouverwill** [comment]
> > The reason I updated these error messages is each one had sent content so made it hard to see error came from. Unique errors do help, and in authorize callback the grant and token validation errors are all caught with same catch so still bit ambiguous as to what happened. What do you think about an error getting thrown in the grant request so you at least know that is what went wrong ?
> 
> I'm sorry I don't follow, the problem is?

ahh sorry, I guess that is two different ideas
1 - the same error message `id_token not present in TokenSet` is used as an error message in several places so it is hard to know the source of the error
2 - inside the `authorizeCallback()` there is a lot of stuff going on with grant(), decryptIdToken() and validateIdToken() which is why I needed to update the grant function so I could see where the failure has happened. Would you be open to consider to having a different error getting thrown if the grant fails?

**Pull request closed without merging**